rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,173 @@
+
+**Kitchen sink** average tachometer benchmark results:
+
+- [**6.7% faster**]: `render-compiled`: 7.93ms - 8.10ms, `render-uncompiled`: 8.50ms - 8.69ms
+- [roughly same]: `update-compiled`: 17.92ms - 18.56ms, `update-uncompiled`: 18.08ms - 19.00ms
+- [roughly same]: `nop-update-compiled`: 5.53ms - 5.66ms, `nop-update-uncompiled`: 5.64ms - 5.80ms
+- [**9.3% larger gzipped file size**]: `index.js` uncompiled is 2216 bytes gzipped, or compiled it is 2443 bytes gzipped.
+
+**Template Heavy** average tachometer benchmark results:
+
+- [**46% faster**]: `render-compiled`: 9.51ms - 9.67ms, `render-uncompiled`: 17.66ms - 18.03ms
+- [**21% faster**]: `update-compiled`: 26.40 - 26.91ms, `update-uncompiled`: 33.62 - 34.51ms
+- [**5.6% larger gzipped file size**]: `template-heavy.js` uncompiled is 202,544 bytes gzipped, or compiled it is 214,495 bytes gzipped.
+
+**repeat** average tachometer benchmark results:
+
+- [**14% faster**]: `render-compiled`: 7.40ms - 7.62ms, `render-uncompiled`: 8.56ms - 8.94ms
+- [**8% faster**]: `update-compiled`: 189.57ms - 197.54ms, `update-uncompiled`: 206.51ms - 218.68ms
+- [**7.1% larger gzipped file size**]: `index.js` uncompiled is 3072 bytes gzipped, or compiled it is 3307 bytes gzipped.
+
+### Interoperability
+
+This labs package is interoperable as it can be safely run on any code, and will only optimize Lit templates.
+
+### Security Impact
+
+The prepared compiled template must be a tagged template literal to defend against JSON injection attacks. Otherwise the `@lit-labs/compiler` package is only a source code transform. Because a compiled template is processed in the same way that an uncompiled template is, we're just moving some of the work to build time. This should preserve the security invariants of the code being compiled.
+
+### Documentation Plan
+
+This package will initially be documented in its own README. If it stays on track to graduation, we should document this package under the _Tools and Workflows_ section on lit.dev. The package on GitHub will also include a TypeScript and JavaScript example of using `@lit-labs/compiler`.
+
+## Downsides
+
+The downside of this approach is that it requires the code to have a build step which can add complexity and build specific setups. There are also slight mismatches in the precompiled prepared HTML, such as lit markers which are not required with pre-compilation. The change in compiled comment markers broke one test in `lit_html` which was introspecting comment markers – something that should be uncommon in production use cases.
+
+## Alternatives
+
+We could not release this package, as the package will have initial implementation cost, and ongoing maintenance cost. The performance wins are large enough that it seems worth it.
+We expect the code to need only occasional maintenance from TypeScript updates, as Lit's template result API is very stable.
```

Comment:
Add a section on what templates are exactly compiled & what have not been compiled, as a reference for the future for which cases we need to handle if we wanted to completely remove the lit-html prepare step.

- Mention raw text elements.
- Svg template
- octal escape
- <template> element with binding.
- Anything else?


In a future RFC we may think about tree shaking lit-html's prepare code, which may require all templates to be compiled.